core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.io.IOException
+import java.util.concurrent.ConcurrentHashMap
+
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.fs.{FileSystem, Path}
+
+import org.apache.spark._
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.{RDD, ReliableCheckpointRDD}
+import org.apache.spark.storage.RDDBlockId
+
+/**
+ * Wrapper around an iterator which writes checkpoint data to HDFS while running action on
+ * an RDD.
+ *
+ * @param id the unique id for a partition of an RDD
+ * @param values the data to be checkpointed
+ * @param fs the FileSystem to use
+ * @param tempOutputPath the temp path to write the checkpoint data
+ * @param finalOutputPath the final path to move the temp file to when finishing checkpointing
+ * @param context the task context
+ * @param blockSize the block size for writing the checkpoint data
+ */
+private[spark] class CheckpointingIterator[T: ClassTag](
+    id: RDDBlockId,
+    values: Iterator[T],
+    fs: FileSystem,
+    tempOutputPath: Path,
+    finalOutputPath: Path,
+    context: TaskContext,
+    blockSize: Int) extends Iterator[T] with Logging {
+
+  private[this] var completed = false
+
+  context.addTaskCompletionListener { ctx =>
+    // We don't know if the task is successful. So it's possible that we still checkpoint the
+    // remaining values even if the task is failed.
+    // TODO optimize the failure case if we can know the task status
+    complete()
+  }
```

Comment:
> also can you put this after the vals?

Just want to make sure `complete()` will run even if any `val` statements fails.